### DISCUSSION TITLE
Possibility of controlling pass before bloom

### DISCUSSION THREAD

**Discussion by VanderSP** [pmndrs/postprocessing#673] [Ideas]
Hi, apologies if this is a naive question or suggestion.

When I implemented selective bloom without using pmndrs tools, I could, for example, replace a mesh's texture with a black-and-white texture during the bloom pass. This allowed me to manually control where the bloom effect was applied, giving me a sort of "manual selective bloom."

In pmndrs, it seems that a mesh is either fully enabled or disabled for bloom. Would it be possible to introduce more granular control, such as changing a texture or material specifically for the bloom pass?

Thank you for your time, and sorry again if this is an obvious or unnecessary question!

**Comment by vanruesc** [comment]
Hi,

changing the texture or material to influence bloom would require an additional render pass for the entire scene. The current selective bloom implementation only needs to render the depth of selected objects again which is then used as a mask. Rendering the scene more than once is one of the main things I'm trying to avoid.

Depth-based selective bloom isn't perfect though: an error threshold is used for the depth-comparison which makes it inaccurate around object edges. This is also the reason why I haven't ported selective bloom to v7 yet.

It's worth mentioning that the basic, non-selective bloom effect already filters bright colors via the luminance threshold. When using high precision framebuffers, colors can exceed the range `[0.0, 1.0]`, so a threshold of `1` can be used to limit bloom to very bright spots. This can be controlled via emissive maps and the emissive intensity to some extent. However, specular highlights and the overall lighting conditions in a scene would still contribute to the absolute luminance which makes this feature more natural, but also more difficult to control.

> Would it be possible to introduce more granular control, such as changing a texture or material specifically for the bloom pass?

To answer the question: there are currently no concrete plans to support this in v7 for performance reasons. But if there's a technique that allows this without much of a performance hit, I'd be happy to add it.

**Comment by VanderSP** [comment]
Very nice, thanks for replying...  makes sense, since performance is crucial on mobile devices!

**Comment by VanderSP** [comment]
gpt said an interesting thing about using stencil buffer, have you ever thought about this? he showed me some examples

**Reply by vanruesc** [reply]
I think this wouldn't be much different from using depth since you'd still need to render the selected objects again to write the stencil bits and this step would also require similar depth testing. So it's not clear how this approach would be an improvement over depth masking.

**Comment by VanderSP** [comment]
Yeah I see where u goin, the coolness is to just shader on shader without multisampling things
